pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,149 @@
+
+// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+// It also returns the namespace of the kubeconfig's current context.
+func (f ClientFactoryFunc) GetClient(kubeconfig []byte) (client.Client, string, error) {
+	return f(kubeconfig)
+}
+
+// DataVolumeManager manages Kubevirt DataVolume operations.
+type DataVolumeManager interface {
+	// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+	CreateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error
+	// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+	// not found object error.
+	GetVolume(ctx context.Context, kubeconfig []byte, name string) (*cdi.DataVolume, error)
+	// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+	ListDataVolumes(ctx context.Context, kubeconfig []byte, listOpts ...client.ListOption) (*cdi.DataVolumeList, error)
+	// DeleteVolume delete the DataVolume based on the passed name and namespace.
+	DeleteVolume(ctx context.Context, kubeconfig []byte, name string) error
+}
+
+type defaultDataVolumeManager struct {
+	client ClientFactory
+	logger logr.Logger
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFactory) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+		logger: log.Log.WithName("kubevirt-data-volume-manager"),
+	}, nil
+}
+
+// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error {
+	c, namespace, err := d.client.GetClient(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	_, err = d.GetVolume(ctx, kubeconfig, dataVolume.Name)
+	if err != nil {
+		if IsDataVolumeNotFoundError(err) {
+			dataVolume.Namespace = namespace
+			return c.Create(ctx, dataVolume)
+		}
+
+		return fmt.Errorf("failed to create data volume: %v", err)
+	}
+
+	return nil
+}
+
+// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+// not found object error.
+func (d *defaultDataVolumeManager) GetVolume(ctx context.Context, kubeconfig []byte, name string) (*cdi.DataVolume, error) {
+	c, namespace, err := d.client.GetClient(kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	dataVolume := &cdi.DataVolume{}
+	if err := c.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, dataVolume); err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil, &DataVolumeNotFoundError{
```

Comment:
I don't think you need to return a special error for `NotFound`, just returning a `nil` data volume should be enough. `nil` should simply always mean "not found".